frontend/ui/widget/imagewidget.lua

```diff
@@ -387,6 +392,82 @@ function ImageWidget:getScaleFactor()
     return self.scale_factor
 end
 
+function ImageWidget:getScaleFactorExtrema()
+    if self._min_scale_factor and self._max_scale_factor then
+        return self._min_scale_factor, self._max_scale_factor
+    end
+
+    -- Compute dynamic limits for the scale factor, based on the screen's area and available memory (if possible).
+    -- Extrema eyeballed to be somewhat sensible given our usual screen dimensions and available RAM.
+    local util = require("util")
+    local memfree, _ = util.calcFreeMem()
+
+    local screen_area = Screen:getWidth() * Screen:getHeight()
+    local min_area = math.ceil(screen_area / 10000)
+    local max_area
+    if memfree then
+        -- If we have access to memory statistics, limit the requested bb size to 25% of the available RAM.
+        local bbtype = self._bb:getType()
+        local bpp
+        if bbtype == Blitbuffer.TYPE_BB8 then
+            bpp = 1
+        elseif bbtype == Blitbuffer.TYPE_BB8A then
+            bpp = 2
+        elseif bbtype == Blitbuffer.TYPE_BBRGB24 then
+            bpp = 3
+        elseif bbtype == Blitbuffer.TYPE_BBRGB32 then
+            bpp = 4
+        elseif bbtype == Blitbuffer.TYPE_BB4 then
+            bpp = 1
+        else
+            bpp = 4
+        end
+
+        max_area = math.floor(0.25 * memfree / bpp)
+    else
+        -- Best effort, trying to account for DPI...
+        local dpi = Screen:getDPI()
+        if dpi <= 212 then
+            max_area = screen_area * 45
+        elseif dpi <= 265 then
+            max_area = screen_area * 25
+        else
+            max_area = screen_area * 15
+        end
+    end
+
+    local area = self._img_w * self._img_h
+    self._min_scale_factor = 1 / math.sqrt(area / min_area)
+    self._max_scale_factor = math.sqrt(max_area / area)
+
+    return self._min_scale_factor, self._max_scale_factor
+end
+
+-- Compute a new best-fit scale factor for a given image rescale percentage
+function ImageWidget:recomputeScaleFactor(scaling)
+    logger.dbg("ImageWidget:recomputeScaleFactor by", scaling)
+    local req_width = self._bb:getWidth() * scaling
+    local req_height = self._bb:getHeight() * scaling
+    logger.dbg("From", self._bb:getWidth(), self._bb:getHeight(), "to", req_width, req_height)
+
+    -- Best fit (in most cases, we're operating on a best-fit scale factor already, so both sides should match).
```

Comment:
Still asking :) "in most cases" - did you meet cases where both sides did not match ?
Also, I dunno about the word "best-fit" : we always keep aspect ratio, we just scale, always keeping the original aspect ration - and we start the image scaled so it's fully contained in the screen, A/R preserved.

Reply:
No, but I haven't check many IV callers ;).

(Best-fit is the wording used in the original code; because there *is* a stretch codepath, but, IIRC, it's only used the the Screensaver and the Screensaver-like plugins).

Reply:
And those caller *should* not trip manual zooms ;).

Reply:
Oh, ok, you thought about these then. Logic splitted between simple ImageViewer and twisted serves-all-use-cases ImageWidget doesn't make things clear :/